src/main/java/org/breedinginsight/services/parsers/excel/ExcelParser.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.breedinginsight.services.parsers.excel;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.poi.ss.usermodel.Cell;
+import org.apache.poi.ss.usermodel.CellType;
+import org.apache.poi.ss.usermodel.Row;
+import org.apache.poi.ss.usermodel.Sheet;
+import org.breedinginsight.services.parsers.ParsingException;
+
+import java.util.*;
+
+/*
+ * This xls/xlsx parser expects spreadsheets with the first row as column names and subsequent rows
+ * as data. Only data in a column with a column name will be kept, data outside these columns is
+ * discarded.
+ */
+@Slf4j
+public class ExcelParser {
+
+    private static final int EXCEL_COLUMN_NAMES_ROW = 0;
+
+    public List<ExcelRecord> parse(Sheet sheet, Set<String> columns) throws ParsingException {
+
+        List<ExcelRecord> records = new ArrayList<>();
+
+        Row columnNames = sheet.getRow(EXCEL_COLUMN_NAMES_ROW);
+
+        if (columnNames == null) {
+            throw new ParsingException("Missing column names row");
+        }
+
+        Map<Integer, String> indexColNameMap = new HashMap<>();
+
+        // get column name to index mapping
+        for(int colIndex=0; colIndex<columnNames.getLastCellNum(); colIndex++) {
+            Cell cell = columnNames.getCell(colIndex, Row.MissingCellPolicy.RETURN_BLANK_AS_NULL);
+            if (cell.getCellType() != CellType.STRING) {
+                throw new ParsingException("Column name must be string cell");
+            }
```

Comment:
If it doesn't already exist, seems like a good unit test to add as well

Reply:
hmm let me see what's going on with that and I'll add a test case to cover it

Reply:
When I delete the trait description column from `data_one_row.csv`, I get a 422 response with the message `Errpr parsing csv: Missing expected columns`. Did you accidentally change something else in the file?

Reply:
Oh, I think you meant you just deleted the value but left the column. In that case I get a 422 response with the message `Missing trait description`

Reply:
Sorry, probably didn't described this well enough. I deleted the value column entirely instead of leaving it blank. So the number of columns in the row didn't match the number of columns in the header.